test/box-luatest/graceful_shutdown_test.lua

```diff
@@ -506,3 +506,18 @@ g.test_graceful_shutdown_not_supported_by_client = function()
     t.assert_equals(tostring(err), 'Peer closed')
     conn:close()
 end
+
+-- Checks that if the net.box callback, which processes box.shutdown event, is
+-- garbage callected while the remote isn't, such partially garbage collected
+-- connection won't block shutdown (gh-7225).
+g.test_net_box_callback_garbage_collected = function()
+    g.server:exec(function()
+        box.ctl.set_on_shutdown_timeout(9000)
+    end)
+    local conn = net.connect(g.server.net_box_uri)
+    t.assert_is_not(conn._callback, nil)
+    conn._callback = nil
```

Comment:
I couldn't understand how is that possible. `conn` is not a weak table. Hence its `_callback` member shouldn't be collected while `conn` has refs. Can you give more details please?

Reply:
The Lua garbage collector runs in step when there's memory pressure, and there's no guarantee it'll perform all steps at once. If the `conn` reference is lost, the garbage collector may collect `conn._callback` and stop before collecting `conn._gc_hook`. It doesn't mean it'll never collect it, but it may take indefinitely long if there's no pressure on Lua memory. As a result, the gc method, which closes the underlying socket won't be called and the server will hang. I added a comment to the test.

Reply:
Thanks for the explanation!